Add room membership endpoints

exul opened the pull request:
- Add the remaining room membership endpoints
- moves `ThirdPartySigned` into the membership module to use it for `join_by_room_id_or_alias` and `join_by_room_id`
- use the `RoomIdOrAlias` enum to allow the user to supply either a room id or an alias, maybe there is a better way?

jimmycuadra commented:
Thanks, exul!

`RoomIdOrAlias` seems like the right approach, but it needs (de)serialization implemented manually, because it needs to serialize into just the ID string without the enum variant wrapper. Similarly, deserialization will be done from a single string, so the deserializer needs to decide which one it is by inspecting sigil rather than relying on an explicit enum variant. I think it makes sense to move that type into ruma-identifiers, if you wanna split that off into another PR over there.

A small nitpick: I prefer the fields of structs to be listed alphabetically.

exul commented:
I updated the PR to use the newly added `RoomIdOrAliasId` that I added to the ruma-identifier crate.
The build currently failes, because the changes for the ruma-identifier crate are not on crates.io yet. I tested it locally by using the github version and it works.
I'll update again once the latest changes for ruma-identifier are on crates.io.

jimmycuadra commented:
Much appreciated!
